### DISCUSSION TITLE
What should we call this repo?

### DISCUSSION THREAD

**Discussion by sndrs** [guardian/csnx#796] [General]
`csnx` is an awkward acronym of _client side Nx_. It has (at least) two big problems:

1. it isn't strictly client side-focussed
2. use of Nx is contingent implementation detail that could easily change

You have to call a repo _something_ and it seemed the least worst option at the time, but now we're considering publishing everything in `/libs` as one mega-package, the problem is back.

A package called `@guardian/csnx` makes a lot of sense, in that it's clear where the code lives, why a feature is included in the package (it's in the repo) and adoption paves the way to moving codebases into the monorepo.

But it's an awful name, for the same reasons above (except _it's not even Nx once it's published_).

So let's find something better. 

- if you have a suggestion, please comment it below
- if you see one you like, please upvote it

**Comment by mxdvl** [comment]
`javascript`

What that ties all of the code in this repo together is that it will be run by a JavaScript runtime, whether in a client or server. It can be confusing that the code itself is written in TypeScript…

**Reply by sndrs** [reply]
ha!

**Reply by sndrs** [reply]
i suggested `js` in a post that took longer to write than yours and missed yours - have removed mine. this was my reasoning:

`https://github.com/guardian/js`

## _all_ js?
it is extremely broad, but then the current description of client side/UI-oriented code also feels naive. 

why shouldn't things like https://github.com/guardian/cdk or https://github.com/guardian/dotcom-rendering live here? 

no one should _have_ to put their code in here, but i think this repo would be doing the right things if it appealed enough that those code owners _want_ to join.

## what about `x`?
supporting non-JS based codebases feels like too big an ambition right now.

## what about `ts`?
we've gone from javascript > flow > typescript, but it's JS at route. e.g. if JS gets types, and we move away from TS, we shouldn't have to create a new repo.

plus not all JS-based code _is_ TS. but no TS-based code is not JS, eventually.

**Reply by jamesmockett** [reply]
Will that always be the case though? The repo is home to Source which may one day support more than just JavaScript. We're already having conversations about design tokens and making the foundation styles available in other formats such as custom properties and variables for iOS and Android.

(_Edit_: Alex's longer reply wasn't there when I wrote the above… 😅)

**Reply by sndrs** [reply]
hmmmt that's a very good counter...going to make another suggestion

**Reply by tjmw** [reply]
I wonder, for non-JS languages, what would the benefit be of keeping that code in this repo/package be? I'm sure this is an oversimplification but from my (limited!) understanding it sounds like the complexity we're trying to tackle here relates to NPM package releases and keeping dependencies in sync? Would that be relevant for non-JS stuff, and is there an argument that that should that be a line in the sand (non-JS doesn't go here?)

**Reply by mxdvl** [reply]
My thinking is that if you’re writing code that will run in a JavaScript runtime, you’ll get the most benefit from being in this monorepo, and this monorepo also exposes an eponymous library on NPM that is to be consumed by other projects that have not yet chosen it as their home.

@jamesmockett – I think that it’s unlikely that the source code of Source will move away from JavaScript, as it’s the language of the web. Whether it exposes tokens in an intermediary format seems more of an implementation detail. It may also be that Source design tokens may come from somewhere else…

**Reply by sndrs** [reply]
it doesn't seem beyond the bounds, though, that if DCR lived here, it would be useful have frontend too. in that case, keeping a play app in a repo called `javascript` seems obviously wrong, even though moving the code there might not be.

**Reply by jamesmockett** [reply]
I agree it's unlikely that the web version of Source is going to move away from JavaScript entirely, but it's possible that we make it more framework agnostic and make greater use of native web technologies rather than React and Emotion. (Or at least provide that as an option. Stylesheets and custom properties for non-React projects for example.) There's also a desire to have an equivalent for native apps so it would make sense that they all live together under the Source umbrella. This almost sounds like an argument for Source going back to being a separate repo which is possibly an entirely different conversation.

**Comment by sndrs** [comment]
`******` (i.e. some non-descriptive name)

given @jamesmockett's [excellent counter argument to `javascript`](https://github.com/guardian/csnx/discussions/796#discussioncomment-6760383), is something non-descriptive that doesn't set bounds a good idea?

I'm generally not in favour of this - it feels like you're not dealing with the fact that you don't understand what you're proposing. 

is that the case here? _should_ we be able to describe what this repo does? 

or is it _actually_ quite nebulous, in a good way?

**Reply by mxdvl** [reply]
Reminds me of https://betterprogramming.pub/software-component-names-should-be-whimsical-and-cryptic-ca260b013de0

**Reply by sndrs** [reply]
I'd not read this, nice one 😃 

This seems like really good advice to me:

>if a name is going to be easily changeable forever [e.g. a variable], please do make it descriptive [...] But if a name is going to serve as, in any sense, an identifier, something that will point at a big complicated thing from many places far away, make it an opaque identifier.

and

>“Descriptive” names don’t create transparency, they create the [illusion of transparency.](https://en.wikipedia.org/wiki/Illusion_of_transparency)

For example, if we wanted to move `frontend` in to a repo called `javascript` it would seem wrong, because it's a play app.

But we should be able to things if _it makes sense to us_, without wondering if we should because the name we gave the repo doesn't describe the new state very well.

I find this convincing...

**Reply by mxdvl** [reply]
[`araneidae`](https://en.wikipedia.org/wiki/Araneidae) 🕸️

**Reply by sndrs** [reply]
maybe a name which is opaque but implies a collection of related things but doesn't describe the relationship?

**Reply by sndrs** [reply]
this was chatgpt's suggestions:

~~1. Batch
2. Stack
3. Heap
4. Array
5. Cluster
6. Bundle
7. Cache
8. Set
9. Pool
10. Pack
11. Group
12. Chain
13. List
14. Block
15. Queue
16. Series
17. Matrix
18. Folder
19. Library
20. Record~~

they are all too loaded/descriptive to my mind

**Reply by sndrs** [reply]
things with good names, off the top of my head: `pasteup`, `source`, ~~`typewriter`~~ `typerighter`, `ophan`

**Reply by sndrs** [reply]
i asked chatgpt to come up with 20 spondees (2 syllables, equally stressed, e.g. ophan) related to collections, number 10 was `haystack` :)

_"i think it's in haystack, good luck finding it though"_

**Reply by jamesmockett** [reply]
The BBC have some wonderfully obscure repo names:
https://github.com/bbc/simorgh

> Named Simorgh after the Persian mythological bird. The Simorgh is the amalgam of many birds (and in some accounts other animals) into one.

**Reply by ParisaTork** [reply]
Off topic, but:

[Simorgh](https://en.wikipedia.org/wiki/Simurgh) comes from the [Shahnameh (The Book of Kings)](https://en.wikipedia.org/wiki/Shahnameh). It is Iran's national epic poem, a seminal work of Persian literature and a very important artifact of Persianate and Middle Eastern history, analogous to the Odyssey and the Iliad, but more wide-reaching in the Eastern world.

One of the cool things about the Shahnameh are the wonderful illustrations. Here's an image of [Zal in Simorgh's nest](https://www.bl.uk/learning/cult/inside/shahnamestories/storyone/birthofzal.html):

<img src="https://github.com/guardian/csnx/assets/47482049/4ffca215-9010-4ca0-929c-6999cbe21909" alt="Zal In Simorgh's Nest" width="300" height="500">

**Fun fact**: I am also named after a Persian mythological creature, called a [Pari](https://en.wikipedia.org/wiki/Peri). And because my dad is my dad, my cousins were called [Sharzad](https://en.wikipedia.org/wiki/Scheherazade) and [Roxanna](https://en.wikipedia.org/wiki/Roxana).

**Comment by arelra** [comment]
`cs`

I don't think 'client-side' necessarily _excludes_ server-side code if we think of 'client-side' work as fullstack.

So it describes an orientation rather than a hard architectural boundary.

**Reply by mxdvl** [reply]
Are you suggesting `fs` / `fullstack`, then? 😉 

---

I wonder if we could spell things out? Is `client-side` too verbose?

**Reply by arelra** [reply]
> fullstack

noooo :)

**Comment by arelra** [comment]
Throwing some more out there:

`rendering`, `gui`, `goo` 

Naming is hard.

**Reply by sndrs** [reply]
we did used to have a `goo` tool, remnants here https://github.com/guardian/frontend/blob/0c1cd1a410bbadd5ad0530250a8149034e23bc77/docs/01-start-here/03-how-to-deploy.md?plain=1#L28.

dcr also began life as `guui` (though no one except me liked it afaict!)

**Comment by sndrs** [comment]
how about `vienna` - it's easy to spell and [tops the quality of life index](https://en.wikipedia.org/wiki/City_quality_of_life_indices)

**Reply by sndrs** [reply]
or `wien` (austrian name)

**Comment by jamesgorrie** [comment]
My suggestion would be to name it what it is for future folk to be able to guess what it does. My next question then is, what is this repo intended for?

(see edit of comment if you want to see me publicly thinking aloud on this)

**Reply by jamesgorrie** [reply]
[A retort](https://betterprogramming.pub/software-component-names-should-be-whimsical-and-cryptic-ca260b013de0) (thanks @sndrs)

**Comment by coldlink** [comment]
`tofu`? A food which can be used in all sorts of flavours and textures. Very woke too.
It's short, and i'm sure you could turn it into an acronym if you really wanted? 
e.g. `(T)ypescript (O)r (F)rontend (U)Is`

**Reply by mxdvl** [reply]
This. A million times this. 🤯 🍜

**Reply by tjmw** [reply]
Love it. I've been trying to think of a recursive acronym we could retrofit a la Gnu, best I've come up with is: `(T)ofu's (O)ur (F)ullstack (U)niverse` 😄

**Reply by raphaelkabo** [reply]
> Love it. I've been trying to think of a recursive acronym we could retrofit a la Gnu, best I've come up with is: `(T)ofu's (O)ur (F)ullstack (U)niverse` 😄


`TOFU Organises Fullstack Usability`?

**Reply by sndrs** [reply]
Types, Objects, Frameworks, Utilities

**Comment by sophie-macmillan** [comment]
I think the issue is that the packages contained in the monorepo are so broad. It currently contains packages that
 - render shared ui (e.g. source-react-components)
 - contains shared config (e.g. prettier and ts-config)
 - contains some sort of other shared functionality (e.g. identity-auth)

That makes it hard to come up with a name thats descriptive enough as to what they all do, without it being too limiting in the future as to what we can include.

So I have a fairly generic suggestion (maybe too generic) of `@guardian/shared` or `@guardian/common`

**Comment by OllysCoding** [comment]
Loads of different and interesting ideas in here so far - but I think the conversation has been quite focussed on _what_ goes in the monorepo, instead of _why_ it goes into the monorepo, therefor I'd like to start the discussion on the idea that a monorepo is usually a series of libs & apps that all depend on each other in various ways - the best reason to put your new lib in CSNX is because if it's using CSNX code, it's just easier.

So, names that are around inter-dependency, linking & sharing seem like the way to go in my opinion (if it's not `tofu` of course)

- `linked`
- `collab`
- `sharer`
- `loops`
- `chainlink`
- `galaxies` (oh dangit, this one is already taken)

Okay so I admit ... I don't _actually_ have any good name ideas for this theme, but I do like the idea of the name being themed around the idea that things go in this repo because they're all sharing with each other, relying on each other, etc.

**Reply by OllysCoding** [reply]
Okay so after some more thinking I've come up with some things I think are actually okay:

- `flock` - in reference to packs of animals who all live & rely on one-another
- `habitat` - focussing on the idea that this is a home for many things which might share an ecosystem (would have to borrow this from the sainsburys subbrand though: https://www.habitat.co.uk/)
- `symbiosis` - bit more self explanatory
- `commune` - living together, helping one-another
- `fungi`

**Reply by mxdvl** [reply]
I like `flock` and `fungi`!

**Reply by jamesmockett** [reply]
I'm not so keen on `fungi` as it has associations with decay which is probably not what we want 😬 

As another submission, `flock` and `commune` reminded me of the name of a nearby art gallery: `collective`

**Reply by OllysCoding** [reply]
I like `collective` too!

**Comment by coldlink** [comment]
Thinking of more journalistic related terms similar to `@guardian/source`

`press` -> like the machine which prints the paper, and this project is "machine" which does that online
`angle` -> from the word meaning the perspective from which the news article is written, and this project provides the `angle` for how code should be used?
`stand` -> like a newsstand, a place where all the projects/code live, like all the papers/magazines on an old newsstand
